winch/codegen/src/masm.rs

```diff
@@ -1257,6 +1269,14 @@ pub(crate) trait MacroAssembler {
     fn mul_wide(&mut self, context: &mut CodeGenContext<Emission>, kind: MulWideKind)
         -> Result<()>;
 
+    /// Takes the int in a `src` operand and replicates it across lanes of
+    /// `size` in a destination result.
+    fn splat_int(&mut self, context: &mut CodeGenContext<Emission>, size: SplatKind) -> Result<()>;
+
+    /// Takes the `src` operand and replicates it across lanes of `size` in
+    /// `dst`.
+    fn splat(&mut self, dst: WritableReg, src: RegImm, size: SplatKind) -> Result<()>;
+
```

Comment:
Taking a look at the callsites, and trying to think to future backends, I'm tempted to suggest adding a single entry here, `splat`, defined as 

```rust
fn splat(&mut self, context: &mut CodeGenContext, kind: SplatKind) {
  ...
}
```

That's potentially the most flexible approach, and that would potentially make the definition more uniform across backends: as far as I can tell, the only reason why we have a special `splat_int` is because of the requirements in x64. 

For that suggestion to work though, we'd probably need to extend `SplatKind` to encode the source type, (e.g., `SplatKind::I8x16`) -- I don't think this changes anything fundamental, since we'd still be able to derive the `lane_size`.

Reply:
The one thing with changing `SplatKind` to encode an `I` or `F` is the `v128.load*_splat` instructions aren't necessarily operating on integers or floats. It's operating on whatever state was stored in the memory. That said, it should still be fine to have the visitor specify a `SplatKind::I32x4` for `v128.load32_splat` even if it's operating on 4 floats, it'll still have the same result.

Reply:
To be explicit, we could rename the current `SplatKind` to `SplatLoadKind` and introduce the types in `SplatKind`. That would tighten the implementation and keep the semantics as close as possible to the Wasm semantics.